Use __HIVE_DEFAULT_PARTITION__ for null partition keys

ulysses-you opened the pull request:
Use` __HIVE_DEFAULT_PARTITION__` for NULL partition keys of all data types.

Before this change, the default value was used only for null partition keys of
type VARCHAR. This happened by accident as these produces empty partition key
which was converted to `__HIVE_DEFAULT_PARTITION__`.

ulysses-you commented:
cc @gggrace14 @mbasmanova do you have time to take a look ? thank you

mbasmanova commented on velox/connectors/hive/HivePartitionUtil.cpp:
```diff
@@ -51,13 +51,23 @@ template <TypeKind Kind>
-};
+  if (partitionVector->as<SimpleVector<T>>()->isNullAt(row)) {
+    return std::make_pair(name, "");
+  } else if (isDate) {
```
nit: remove else after return

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +125,51 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+    std::iota(partitionChannels.begin(), partitionChannels.end(), 0);
+
+    EXPECT_EQ(
+        FileUtils::makePartName(extractPartitionKeyValues(
```
This change modifies extractPartitionKeyValues function, hence it would be nice to test just that function and its return value. FileUtils::makePartName can be tested separately to ensure that it converts empty value into __HIVE_DEFAULT_PARTITION__.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +125,51 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+           std::vector<std::optional<int32_t>>{std::nullopt}),
+       makeNullableFlatVector<int64_t>(
+           std::vector<std::optional<int64_t>>{std::nullopt}),
+       makeNullableDictionary<StringView>(
```
The usage of makeNullableDictionary here is strange. It seems to create a vector with 10 rows while other columns have only 1 row. Is this intentional? 

I assume makeNullableDictionary function is not needed and can be replaced with makeNullableFlatVector.

mbasmanova reviewed (commented):
@ulysses-you The changes look file % some comments. Curious, how did you run into this? Are you having some use case which generates partitioned with null partition key values? Does Spark support this?

ulysses-you commented:
thank you @mbasmanova , addressed the comments.

yes, when we support dynamiac partition write for Gluten, some Spark tests failed due to this issue. e.g., 
`INSERT INTO TABLE t PARTITION(p) INSERT c1, p FROM t`, the `p` is a nullable int type.
Spark will cast all types to string first so if the value is null then the partition path name will be `__HIVE_DEFAULT_PARTITION__ `. The related Spark code is [here](
https://github.com/apache/spark/blob/7f056d89164c584ec57e252eb37bdc17a0a2e20c/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L221-L228)

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,66 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+           std::vector<std::optional<StringView>>{std::nullopt}),
+       makeConstant<int32_t>(std::nullopt, 1, DATE())});
+
+  std::vector<std::pair<std::string, std::string>> expectedPartitionKeyValues{
```
Since all values are empty, this can be just a vector of first values

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,66 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+      std::make_pair("dict_string_col", ""),
+      std::make_pair("const_date_col", "")};
+
+  std::vector<std::string> expectedPartitionNames;
```
This is not needed. Just compute expectedPartitionName inside the loop.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,66 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+        return p.first + "=__HIVE_DEFAULT_PARTITION__";
+      });
+
+  std::vector<column_index_t> partitionChannels;
```
simplify the code by moving this variable inside the loop:

std::vector<column_index_t> partitionChannels(i);

ulysses-you commented:
thank you @mbasmanova , addressed the comments.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,47 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+    for (const auto& entry : partitionEntries) {
+      EXPECT_EQ(entry.second, "");
+    }
+    auto expectedPartitionNames = folly::join(
```
expectedPartitionNames -> expectedPartitionName (no 's' at the end)

Why do you feel the need to test makePartName here? Don't we have a separate unit test for makePartName to verify that it converts empty values to __HIVE_DEFAULT_PARTITION__?

Anyway, this test only verifies the case when all partition key values are null, but doesn't test cases where some are null and others are not.

ulysses-you replied:
I did not realize we have a separate unit test for makePartName. I agree it's unnecessary, removed.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,47 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+      "flat_smallint_col",
+      "flat_int_col",
+      "flat_bigint_col",
+      "dict_string_col",
```
dict -> flat

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,47 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+           std::vector<std::optional<StringView>>{std::nullopt}),
+       makeConstant<int32_t>(std::nullopt, 1, DATE())});
+
+  for (auto i = 1; i <= partitionColumnNames.size(); i++) {
```
What's the point of testing 1, 2, 3, ...N partition keys? Why not just test all partition key types, but use single partition key for each test?

ulysses-you replied:
it follows previous test, I changed to use single partition key for each test.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,47 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+  RowVectorPtr input = makeRowVector(
+      partitionColumnNames,
+      {makeNullableFlatVector<bool>(
+           std::vector<std::optional<bool>>{std::nullopt}),
```
Can you drop 'std::vector<std::optional<bool>>'?

ulysses-you replied:
sure, removed

mbasmanova reviewed (commented):
Looks good, but I still have some questions about the test. Overall, it is not very clear what's the thinking behind the test design. Would you clarify?

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -117,3 +117,33 @@ TEST_F(HivePartitionUtilTest, partitionName) {
+    std::vector<column_index_t> partitionChannels = {(column_index_t)i};
+    auto partitionEntries = extractPartitionKeyValues(
+        makePartitionsVector(input, partitionChannels), 0);
+    for (const auto& entry : partitionEntries) {
```
This loop is unnecessary since there is only 1 value. It is more robust to assert number of entries in partitionEntries and then check the only entry. It is better to check both first and second values.

```
EXPECT_EQ(1, partitionEntries.size());
EXPECT_EQ(partitionColumnNames[i], partitionEntries[0].first);
EXPECT_EQ("", partitionEntries[0].second);
```

ulysses-you replied:
+1, addressed

mbasmanova reviewed (commented):
Looks good % one nit. Thanks.

mbasmanova reviewed (approved):
Thanks.
